api/src/controllers/AuthController.js

```diff
@@ -161,13 +179,51 @@ const jwt = (req, res, next) => {
     user.authFailedAttempts = 0;
     user.authLastAttempt = new Date();
     return user.save(() =>
-       createUserJWT(user).then()
-          .then(token => res.send(token))
-          .catch(authFailure)
+      Promise.all([createUserJWT(user), createUserRefreshJWT(user)])
+        .then(
+          ([accessToken, refreshToken]) => res.cookie(
+            `refresh_token_user_${user._id}`,
+            refreshToken,
+            buildRefreshCookieOption(req.protocol),
+          ).send(accessToken))
+        .catch(authFailure)
     );
   })(req, res, next);
 };
 
+const jwtRefresh = (req, res, next) => {
+  (async () => {
+    try {
+      const refreshToken = req.cookies[`refresh_token_${req.body.tokenType}_${req.body.id}`];
+      const decodedToken = await jsonwebtoken.verify(refreshToken, process.env.APP_SECRET);
+
+      const { tokenId, tokenType, userId, provider } = decodedToken;
+
+      const user = await User.findOne({ _id: userId });
+
+      if (!user) {
+        throw new Unauthorized('No User');
+      }
+
+      if (tokenType === 'user_refresh') {
+        const newUserToken = await createUserJWT(user, provider);
+        res.status(200).send(newUserToken);
+      } else if (tokenType === 'organisation_refresh') {
+        const newOrgToken = await createOrgJWT(user, tokenId, provider);
+        res.status(200).send(newOrgToken);
+      } else {
+        throw new Unauthorized('Invalid tokenType');
+      }
+    } catch (err) {
+      if (!['JsonWebTokenError', 'TokenExpiredError'].includes(err.name) && !(err instanceof Unauthorized)) {
+        console.error(err);
+      }
+      res.status(401).send('Unauthorized');
+    }
+  })().catch(next);
```

Comment:
I'm a little uncertain about what happens when you call next with an error like Unauthorised. Could you please clarify what happens in these cases @h-kanazawa? Typically we would just handle all errors in the route handler to avoid having to read Express docs and trace down any code that might handle this exception elsewhere.

Reply:
I think we'd be back to an unhandled exception with a 500 thrown by Express. Might be nicer to wrap in [`catchErrors`](https://github.com/LearningLocker/learninglocker/blob/a3e3ab4290f22d579f6e908b38231435b4a57bf4/api/src/controllers/ExportController.js#L13) as we do elsewhere, but unsure if that presents issues

Reply:
Sorry. This async anonymous function ignores errors. So `.catch` will never be called. Furthermore, `jwtRefresh` doesn't need `next` function. I rewrite this with catchErrors.